x/auth/ante.go

```diff
@@ -27,10 +27,14 @@ func init() {
 	copy(simSecp256k1Pubkey[:], bz)
 }
 
+// SignatureVerificationGasConsumer is the type of function that is used to both consume gas when verifying signatures
+// and also to accept or reject different types of PubKey's. This is where apps can define their own PubKey types.
+type SignatureVerificationGasConsumer = func(meter sdk.GasMeter, sig []byte, pubkey crypto.PubKey, params Params) sdk.Result
```

Comment:
Can you provide an example of an alternative implementation to the default? I'm trying to see where the motivation for this stems from.

Reply:
So the motivation is to let SDK users configure the keys their zone uses. As far as I can tell the sig verification gas consumer is the only function that accepts/rejects different key types.

Here's an example that adds a new "fancynewcryptoalg" key type and also uses concrete pub key types rather than the string comparison in the current default implementation:
```go
func AltSigVerificationGasConsumer(
	meter sdk.GasMeter, sig []byte, pubkey crypto.PubKey, params Params,
) sdk.Result {
	switch pubkey := pubkey.(type) {
        case fancynewcryptoalg.PubKeyFancyNewCryptoAlg
		meter.ConsumeGas(params.SigVerifyCostSecp256k1, "ante verify: fancyNewCryptoAlg")
		return sdk.Result{}

	case ed25519.PubKeyEd25519:
		meter.ConsumeGas(params.SigVerifyCostED25519, "ante verify: ed25519")
		return sdk.Result{}

	case secp256k1.PubKeySecp256k1:
		meter.ConsumeGas(params.SigVerifyCostSecp256k1, "ante verify: secp256k1")
		return sdk.Result{}

	
	case multisig.PubKeyMultisigThreshold:
		var multisignature multisig.Multisignature
		codec.Cdc.MustUnmarshalBinaryBare(sig, &multisignature)

		consumeMultisignatureVerificationGas(meter, multisignature, pubkey, params)
		return sdk.Result{}

	default:
		return sdk.ErrInvalidPubKey(fmt.Sprintf("unrecognized public key type: %s", pubkey)).Result()
	}
}
```

Reply:
Neat! Also, keep in mind, the original implementation also type switched. But there were reasons due to Ethermint where I moved it to a more...hacky...string-based comparison. Ideally, the PubKey type would have a `Type()` method on the interface.

Reply:
So if you had a configurable sig verifier like I'm proposing, Ethermint could just have it's own implementation that type switched and we wouldn't need the hacky string comparison in the default.

Reply:
Yes, precisely!